### ISSUE TITLE
Is there a way to run a piece of code when an assertion fails inside an end block?

### ISSUE THREAD

**Issue by asdf01** [chaijs/chai#755]
I find myself doing this in my tests:
chai.request(app)
  .put('/user/me')
  .send({ password: '123', confirmPassword: '123' })
  .end(function (err, res) {
     **try {**
       expect(err).to.be.null;
       expect(res).to.have.status(200);
     **} catch (e) {
       //write err and res objects to custom log file
       throw e;
     }**
  });
I add try and catch blocks around the chai assertion checking code so that I can log out the error and response objects to help me debug the issue.  Is there another way to do this?  Thanks.

**Comment by keithamus** [comment]
Hey @asdf01 thanks for the issue.

Looks like this is a chai-http issue, but that's okay. I'd recommend you have a look at using Promises instead, as it'll make life much easier:

```js
chai.request(app)
  .put('/user/me')
  .send({ password: '123', confirmPassword: '123' })
  .then((res) => {
    expect(res).to.have.status(200);
  })
  .catch((e) => {
    //write err and res objects to custom log file
    throw e
  })
```

The `then` will only execute if there were no errors, the `catch` will catch any errors from the request, _as well as any_ errors from the `then` function, giving you a chance to log errors out.


Of course, if you're using this in a test harness, something like Mocha, then you can use the error handling mechanisms from within that tool. For example:

```js
it('returns 200 OK', (done) => {

  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    })
    .then(done, done)

});
```

Mocha's `done` callback takes 1 argument, an error. The second `.then` is passed both the fulfilled handler (the first `done`) which will be given no arguments (as the first `.then` doesn't return anything), and the second `done` is the rejection handler which will be given the error object. Meaning this just works™.

If you're using Mocha you can also make use of its support for Promises out of the box, and you can return the Promise from within the test (like I've done below) and it'll add the `done` stuff for you automatically:

```js
it('returns 200 OK', () => {

  return chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    });

});
```

You could also couple this with [chai-as-promised](https://github.com/domenic/chai-as-promised/) to do some powerful inline assertions, for example:

```js
it('returns 200 OK', () => {

  return expect(chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
  ).to.eventually.have.status(200);

});
```

So there you have it. Much cleaner code, using Promises!

I'll close this issue, as hopefully this answers your question. If you have any more questions feel free to continue discussing here, or open a new issue if you have one. Also, in future, if you have general questions like this - they may be better asked on a site like [StackOverflow](stackoverflow.com)

**Comment by asdf01** [comment]
Hi @keithamus 

Thanks for looking into this question.  You have made the whole process very pleasant.  My past experience of asking questions on the internet usually results in receiving tirades of abuse and ends with me questioning humanity.  But you have made this process surprisingly pleasant, so thank you. 

Your response has made me realise that those things are all JS Promises and can be chained together.  So it has made me think about these problems in a different way.  I tried out your suggestions, I don't want to sound ungrateful but your suggestions don't really allow me to achieve what I need to do, which is to log the err and the res objects from the http response when an assertion fails.  

About this suggestion:
```
chai.request(app)
  .put('/user/me')
  .send({ password: '123', confirmPassword: '123' })
  .then((res) => {
    expect(res).to.have.status(200);
  })
  .catch((e) => {
    //write err and res objects to custom log file
    throw e
  })
```
Yes I am using chai-http inside mocha.  In the mocha context, the following problems exists with the above approach:
- we no longer have access to the err object anywhere
- inside the catch, I don't have access to the err and res objects to write to the logs
- inside the catch, despite us calling `throw e`, mocha no longer ends the test, rather it just times out.  But if we call `done()` from within the catch, it results in a successful test outcome, which is not the outcome that we want to convey to mocha.

About this suggestion:
```
it('returns 200 OK', (done) => {
  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    })
    .then(done, done)
});
```
I was a bit confused by this suggestion.  I'm guessing you meant something like this:
```
it('returns 200 OK', (done) => {
  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    }).then(done, function(e){
      //write err and res objects to custom log file
      throw e;
    });
});
```
If so, then a similar problem exits for this approach:
- we no longer have access to the err object
- in the error handler of the 2nd then: it does not have access to the err and the res object when an assertion fails
- in the error handler of the 2nd then: `throw e` results in a timeout.  `done()` results in a successful test, which is not what we want to convey to mocha.

About this suggestion:
```
it('returns 200 OK', () => {
  return chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    });
});
```
I have the following issues:
- it only outputs to the console.  For large numbers of tests it's easier for me to be able to look through a log file which is why I used a custom logger
- mocha doesn't log enough information about the error.  When I do my custom error handing I use nodejs' util.inspect(res) to log out the full res and err objects. 

About this suggestion:
```
it('returns 200 OK', () => {
  return expect(chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
  ).to.eventually.have.status(200);
});
```
I didn't try this as it looks like it only allows me to assert one thing.  I usually want to assert more than one thing in a test.

Again I would like to thank you for taking the time to look into this question.  Do you have any other ideas?

**Comment by keithamus** [comment]
> we no longer have access to the err object anywhere

The error object is the `e` in the `catch`.

> inside the catch, I don't have access to the err and res objects to write to the logs

With chai http, the response object is attached to the error. A small exaple:

```js
chai.request(app)
  .put('/user/me')
  .send({ password: '123', confirmPassword: '123' })
  .then((response) => {
    console.log('status is ', response.status);
  })
  .catch((error) => {
    console.log('status is ', error.response.status);
  })
```

> inside the catch, despite us calling throw e, mocha no longer ends the test, rather it just times out. But if we call done() from within the catch, it results in a successful test outcome, which is not the outcome that we want to convey to mocha.

Yup, in Mocha you need to either return the promise, or call `done`. In my example:

```js
it('returns 200 OK', (done) => {
  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    })
    .then(done, done)
});
```

The `.then` is given `done` twice for good reason. If we expand this it might make more sense:

```js
it('returns 200 OK', (done) => {
  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
      return undefined;
    })
    .then((result) => done(result /* undefined */), (error) => done(error /* Error */))
});
```

You can see that the first `done` is called with `undefined` and so Mocha considers that a test pass. The second `done` is called with `error` and Mocha sees this as a test fail. If you want do do stuff in the catch, you totally can but you always need to call done:

```js
it('returns 200 OK', (done) => {
  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    })
    .catch((error) => {
      //write err and res objects to custom log file
     throw error;
    })
    .then(done, done);
});

// or you can do this:

it('returns 200 OK', (done) => {
  chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
      done();
    })
    .catch((error) => {
      //write err and res objects to custom log file
     done(error);
   });
});
```

> mocha doesn't log enough information about the error. When I do my custom error handing I use nodejs' util.inspect(res) to log out the full res and err objects.

If mocha doesn't log enough, you can always mix and match as you see fit. For example:

```js
it('returns 200 OK', () => {
  return chai.request(app)
    .put('/user/me')
    .send({ password: '123', confirmPassword: '123' })
    .then((res) => {
      expect(res).to.have.status(200);
    })
    .catch((error) => {
      // write err and res objects to custom log file
      throw e;
    });
});
```

As long as you always throw the error, the Promise will reject and Mocha can work with it.

---

To summarise: with Promises you can add as many `.then` or `.catch` commands as you want. If you want the whole chain to fail, then make sure you rethrow errors in your `catch`.

**Comment by asdf01** [comment]
Hi @keithamus 

Thanks for responding so quickly to the issue.  Sorry for my late reply, I was pulled to another part of our project.  I'm responding to show you that I value the time you have spent on this issue for me.  But I hate to sound argumentative but some of the suggestions just don't seem to work.  

>> inside the catch, I don't have access to the err and res objects to write to the logs

> With chai http, the response object is attached to the error

Testing this example again:
```
chai.request(app)
  .put('/user/me')
  .send({ password: '123', confirmPassword: '123' })
  .then((response) => {
    console.log('status is ', response.status);
  })
  .catch((error) => {
    console.log('status is ', error.response.status);
  })
```
If an exception or an assertion failure occurs in the then block, the catch block contents is never called.  
I don't know when it's structured like this, what type of error it catches, maybe when something javascripty goes wrong when making the request process or if the request times out maybe.  

When it's structured in this way that you suggested:
```
    chai.request(server)
      .put('/user/me')
      .then((response) => {
        console.log('response status: ', response.status);
        response.should.have.status(200);
      })
      .then(done, (error) => {
        console.log("error: " + util.inspect(error));
        done(error);
      })
```
* yes the error handler does get called when an assertion fails
* but the error object seems to be about the assertion failure and no longer has the original http response, which is what i'm after

> The .then is given done twice for good reason

I think I understand what the second done is for now, thanks.

> As long as you always throw the error, the Promise will reject and Mocha can work with it.

I couldn't use the example you provided because when I ran it the catch block would never get called for the reasons I suggested earlier.  So I tried something like this:
```
    chai.request(server)
      .put('/user/me')
      .then((response) => {
        console.log('success: ', response.status);
        try {
          response.should.have.status(200);
        } catch(error) {
          console.log("catch called");
          throw error;
        }
      })
```
In the scenario above, in the context of mocha, the **throw error** seems to do nothing, mocha just times out.  When I replace the **throw error** with **done(error)** only then does mocha become aware of the error.  

>> I add try and catch blocks around the chai assertion checking code so that I can log out the error and response objects to help me debug a test failure

At this point I would like to thank you again for going on this journey with me.  But looking back at the original goal, I need to be able to write the entire context of the test out to a log file when a test fails.  Apart from adding ugly try catch blocks, the only way I have been able to achieve this is by storing the err and res values into a temporary variable and clearing them in mocha's beforeEach and just before the done() call.  The afterEach function would log those objects out if those objects haven't been cleared.  This approach isn't slick but it's a lot less code than the try catch approach.  

The ideal outcome would be if this was a configurable / switchable feature in chai-http.  I'm guessing it might be possible since the err and res objects come from chai-http is that right?

Hopefully I haven't missed anything in your suggestions.  I hope you understand I'm not trying to be argumentative and I'm merely trying to provide you feedback about my observations since you've so kindly donated your time into trying to help me solve my problems.  Thanks again for contributing to this awe inspiring world of open source stuff, without people like you, people like me probably won't have jobs.  Thanks.

**Comment by keithamus** [comment]
> If an exception or an assertion failure occurs in the then block, the catch block contents is never called.

Not sure why this is failing for you. This is a fundamental tenant of Promises. Catch blocks catch all `then` errors before them. To wit:

```js
Promise.resolve(4)
  .then((value) => { throw new Error(`got ${value} but still erroring`); }
  .catch((error) => console.log(error));
```

Running this in the Node repl I get:

```js
> Error: got 4 but still erroring
    at Promise.resolve.then (repl:1:44)
    at process._tickDomainCallback (internal/process/next_tick.js:129:7)
```

I suppose if this doesn't work for you, and you can reliably reproduce it, then we may have a bug. Otherwise I would suggest there might be something going in your code which is causing problems.

**Comment by meeber** [comment]
> If an exception or an assertion failure occurs in the then block, the catch block contents is never called.

I think what's probably happening is that the `catch` block is being called, but a second error is occurring inside the `catch` block. This second error is silently swallowed, preventing the rest of the `catch` from running to completion, and causing the test to timeout.

Consider the following example:

```js
it("meep", done => {
  chai.request("http://www.google.com")
    .get("/")
    .then(res => {
      console.log("'then' block res.status is ", res.status);
      expect(res).to.have.status(999);
    })
    .catch(err => {
      console.log("'catch' block err is ", ert);
      done(err);
    })
});
```

This example outputs:

> 'then' block res.status is  200
>   1) meep
> 
>   0 passing (2s)
>   1 failing
> 
>   1)  meep:
>      Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.

At first glance, the issue appears to be that the `catch` block isn't being executed. In actuality, the issue is that I misspelled `err` as `ert`, causing a second error to be thrown and swallowed inside the `catch` block. Fix that typo and the output is now:

> 'then' block res.status is  200
> 'err' block err is  { AssertionError: expected { Object (domain, _events, ...) } to have status code 999 but got 200
>     at chai.request.get.then.res (/opt/code/meep/test/index.js:13:27)
>     at process._tickCallback (internal/process/next_tick.js:103:7)
>   message: 'expected { Object (domain, _events, ...) } to have status code 999 but got 200',
>   showDiff: false,
>   actual: 200,
>   expected: 999 }
>   1) meep
> 
>   0 passing (241ms)
>   1 failing
> 
>   1)  meep:
>      AssertionError: expected { Object (domain, _events, ...) } to have status code 999 but got 200
>       at chai.request.get.then.res (test/index.js:13:27)
>       at process._tickCallback (internal/process/next_tick.js:103:7)

**Wooo, `catch` block runs to completion. But unfortunately this doesn't bring us much closer to solving the central issue:**

> but the error object seems to be about the assertion failure and no longer has the original http response, which is what i'm after

You're right, the Chai Assertion Error exposes the expected and actual values, but it doesn't provide the original response object. Personally I'd do something like this:

```js
it("meep", () => {
  let res = "_PRETEST";

  return chai.request("http://www.google.com")
    .get("/")
    .then(_res => {
      res = _res;
      expect(res).to.have.status(999);
    })
    .catch(err => {
      console.log("err is ", err);
      console.log("res is ", res);
    });
});
```

This strategy catches and logs the error regardless if the error is caused by the request, or by the chai assertion(s) performed on the response object. It also logs either the response object, or "_PRETEST" if the response object never gets set. An added benefit is that you'll be aware if the request call for some reason explicitly set `res` to `undefined`, a point that might otherwise go by undetected if `res` started out as `undefined` instead of as "_PRETEST".

**Comment by PSeitz** [comment]
If I want to convert a normal assert to a Promise, to add logging in a error case, is there something more elegant that this:



       var differences = deep.diff(lhs, rhs) 
       return Promise.resolve(4).then(() => {
            return expect(differences).to.equal(undefined)
        })
        .catch((error) => {
            console.log(differences)
            throw error
        })

**Comment by meeber** [comment]
If you're not working with promises or asynchronicity already like in the original poster's example, then there's no need to introduce them into the mix. Just use a `try`/`catch` block:

```js
var differences = deep.diff(lhs, rhs)
try {
  expect(differences).to.equal(undefined)
} catch (err) {
  console.log(differences)
  throw err
}
```
